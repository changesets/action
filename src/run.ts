import {
  execWithOutput,
  extractPublishedPackages,
  requireChangesetsCliPkgJson,
} from "./utils";
import resolveFrom from "resolve-from";

type PublishOptions = {
  tagName: string;
  cwd?: string;
};

export type PublishedPackage = { name: string; version: string };

export type PublishResult =
  | {
      published: true;
      publishedPackages: PublishedPackage[];
    }
  | {
      published: false;
    };

export async function runVersion({
  tagName,
  cwd = process.cwd(),
}: PublishOptions) {
  requireChangesetsCliPkgJson(cwd);
  console.info(`Running version workflow from cwd:`, cwd);

  let changesetVersionOutput = await execWithOutput(
    "node",
    [
      resolveFrom(cwd, "@changesets/cli/bin.js"),
      "version",
      "--snapshot",
      tagName,
    ],
    {
      cwd,
    }
  );

  if (changesetVersionOutput.code !== 0) {
    throw new Error(
      "Changeset command exited with non-zero code. Please check the output and fix the issue."
    );
  }
}

export async function runPublish({
  tagName,
  cwd = process.cwd(),
}: PublishOptions): Promise<PublishResult> {
  requireChangesetsCliPkgJson(cwd);
  console.info(`Running publish workflow...`);

  let changesetPublishOutput = await execWithOutput(
    "node",
    [
      resolveFrom(cwd, "@changesets/cli/bin.js"),
      "publish",
      "--no-git-tag",
      "--tag",
      tagName,
    ],
    {
      cwd,
    }
  );

  if (changesetPublishOutput.code !== 0) {
    throw new Error(
      "Changeset command exited with non-zero code. Please check the output and fix the issue."
    );
  }

  let releasedPackages: PublishedPackage[] = [];

  for (let line of changesetPublishOutput.stdout.split("\n")) {
    let match = extractPublishedPackages(line);

    if (match === null) {
      continue;
    }

    releasedPackages.push(match);
  }

  const publishedAsString = releasedPackages
    .map((t) => `${t.name}@${t.version}`)
    .join("\n");

  const released = releasedPackages.length > 0;

  if (released) {
    console.info(
      `Published the following pakages (total of ${releasedPackages.length}): ${publishedAsString}`
    );
  } else {
    console.info(`No packages were published...`);
  }

  if (releasedPackages.length) {
    return {
      published: true,
      publishedPackages: releasedPackages,
    };
  }

  return { published: false };
}
