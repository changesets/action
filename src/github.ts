import * as github from "@actions/github";

import { PublishedPackage, PublishResult } from "./run";

const SNAPSHOT_COMMENT_IDENTIFIER = `<!-- changesetsSnapshotPrCommentKey -->`;

function formatTable(packages: PublishedPackage[]): string {
  const header = `| Package | Version | Info |\n|------|---------|----|`;

  return `${header}\n${packages
    .map(
      (t) =>
        `| \`${t.name}\` | \`${t.version}\` | [npm â†—ï¸Ž](https://www.npmjs.com/package/${t.name}/v/${t.version}) [unpkg â†—ï¸Ž](https://unpkg.com/browse/${t.name}@${t.version}/) |`
    )
    .join("\n")}`;
}

export async function upsertComment(options: {
  tagName: string;
  token: string;
  publishResult: PublishResult;
}) {
  const octokit = github.getOctokit(options.token);
  const issueContext = github.context.issue;

  if (!issueContext?.number) {
    console.log(
      `Failed to locate a PR associated with the Action context, skipping Snapshot info comment...`
    );
  }

  let commentBody =
    options.publishResult.published === true
      ? `### ðŸš€ Snapshot Release (\`${
          options.tagName
        }\`)\n\nThe latest changes of this PR are available as \`${
          options.tagName
        }\` on npm (based on the declared \`changesets\`):\n${formatTable(
          options.publishResult.publishedPackages
        )}`
      : `The latest changes of this PR are not available as \`${options.tagName}\`, since there are no linked \`changesets\` for this PR.`;

  commentBody = `${SNAPSHOT_COMMENT_IDENTIFIER}\n${commentBody}`;

  const existingComments = await octokit.rest.issues.listComments({
    ...github.context.repo,
    issue_number: issueContext.number,
    per_page: 100,
  });

  const existingComment = existingComments.data.find((v) =>
    v.body?.startsWith(SNAPSHOT_COMMENT_IDENTIFIER)
  );

  if (existingComment) {
    console.info(
      `Found an existing comment, doing a comment update...`,
      existingComment
    );

    const response = await octokit.rest.issues.updateComment({
      ...github.context.repo,
      body: commentBody,
      comment_id: existingComment.id,
    });

    console.log(`GitHub API response:`, response.status);
  } else {
    console.info(`Did not found an existing comment, creating comment..`);

    const response = await octokit.rest.issues.createComment({
      ...github.context.repo,
      body: commentBody,
      issue_number: issueContext.number,
    });

    console.log(`GitHub API response:`, response.status);
  }
}
