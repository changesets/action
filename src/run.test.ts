import fixturez from "fixturez";
import * as github from "@actions/github";
import fs from "fs-extra";
import path from "path";
import writeChangeset from "@changesets/write";
import { Changeset } from "@changesets/types";
import { runVersion } from "./run";

jest.mock("@actions/github", () => ({
  context: {
    repo: {
      owner: "changesets",
      repo: "action",
    },
    ref: "refs/heads/some-branch",
    sha: "xeac7",
  },
  getOctokit: jest.fn(),
}));
jest.mock("./gitUtils");

let mockedGithubMethods = {
  search: {
    issuesAndPullRequests: jest.fn(),
  },
  pulls: {
    create: jest.fn(),
  },
  repos: {
    createRelease: jest.fn(),
  },
};
(github.getOctokit as any).mockImplementation(() => mockedGithubMethods);

let f = fixturez(__dirname);

const linkNodeModules = async (cwd: string) => {
  await fs.symlink(
    path.join(__dirname, "..", "node_modules"),
    path.join(cwd, "node_modules")
  );
};
const writeChangesets = (changesets: Changeset[], cwd: string) => {
  return Promise.all(changesets.map((commit) => writeChangeset(commit, cwd)));
};

beforeEach(() => {
  jest.clearAllMocks();
});

describe("version", () => {
  it("creates simple PR", async () => {
    let cwd = f.copy("simple-project");
    linkNodeModules(cwd);

    mockedGithubMethods.search.issuesAndPullRequests.mockImplementationOnce(
      () => ({ data: { items: [] } })
    );

    mockedGithubMethods.pulls.create.mockImplementationOnce(
      () => ({ data: { number: 123 } })
    );

    await writeChangesets(
      [
        {
          releases: [
            {
              name: "simple-project-pkg-a",
              type: "minor",
            },
            {
              name: "simple-project-pkg-b",
              type: "minor",
            },
          ],
          summary: "Awesome feature",
        },
      ],
      cwd
    );

    await runVersion({
      githubToken: "@@GITHUB_TOKEN",
      cwd,
    });

    expect(mockedGithubMethods.pulls.create.mock.calls[0]).toMatchSnapshot();
  });

  it("only includes bumped packages in the PR body", async () => {
    let cwd = f.copy("simple-project");
    linkNodeModules(cwd);

    mockedGithubMethods.search.issuesAndPullRequests.mockImplementationOnce(
      () => ({ data: { items: [] } })
    );

    mockedGithubMethods.pulls.create.mockImplementationOnce(
      () => ({ data: { number: 123 } })
    );

    await writeChangesets(
      [
        {
          releases: [
            {
              name: "simple-project-pkg-a",
              type: "minor",
            },
          ],
          summary: "Awesome feature",
        },
      ],
      cwd
    );

    await runVersion({
      githubToken: "@@GITHUB_TOKEN",
      cwd,
    });

    expect(mockedGithubMethods.pulls.create.mock.calls[0]).toMatchSnapshot();
  });

  it("doesn't include ignored package that got a dependency update in the PR body", async () => {
    let cwd = f.copy("ignored-package");
    linkNodeModules(cwd);

    mockedGithubMethods.search.issuesAndPullRequests.mockImplementationOnce(
      () => ({ data: { items: [] } })
    );

    mockedGithubMethods.pulls.create.mockImplementationOnce(
      () => ({ data: { number: 123 } })
    );

    await writeChangesets(
      [
        {
          releases: [
            {
              name: "ignored-package-pkg-b",
              type: "minor",
            },
          ],
          summary: "Awesome feature",
        },
      ],
      cwd
    );

    await runVersion({
      githubToken: "@@GITHUB_TOKEN",
      cwd,
    });

    expect(mockedGithubMethods.pulls.create.mock.calls[0]).toMatchSnapshot();
  });
});
