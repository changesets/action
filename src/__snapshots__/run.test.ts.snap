// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version creates simple PR 1`] = `
[
  {
    "base": "some-branch",
    "body": "This PR was opened by the [Changesets release](https://github.com/changesets/action) GitHub action. When you're ready to do a release, you can merge this and publish to npm yourself or [setup this action to publish automatically](https://github.com/changesets/action#with-publishing). If you're not ready to do a release yet, that's fine, whenever you add more changesets to some-branch, this PR will be updated.


# Releases
## simple-project-pkg-a@1.1.0

### Minor Changes

-   Awesome feature

### Patch Changes

-   Updated dependencies
    -   simple-project-pkg-b@1.1.0

## simple-project-pkg-b@1.1.0

### Minor Changes

-   Awesome feature
",
    "head": "changeset-release/some-branch",
    "owner": "changesets",
    "repo": "action",
    "title": "Version Packages",
  },
]
`;

exports[`version does not include any release information if a message with simplified release info exceeds size limit 1`] = `
[
  {
    "base": "some-branch",
    "body": "This PR was opened by the [Changesets release](https://github.com/changesets/action) GitHub action. When you're ready to do a release, you can merge this and publish to npm yourself or [setup this action to publish automatically](https://github.com/changesets/action#with-publishing). If you're not ready to do a release yet, that's fine, whenever you add more changesets to some-branch, this PR will be updated.


# Releases

> All release information have been omitted from this message, as the content exceeds the size limit.",
    "head": "changeset-release/some-branch",
    "owner": "changesets",
    "repo": "action",
    "title": "Version Packages",
  },
]
`;

exports[`version does not include changelog entries if full message exceeds size limit 1`] = `
[
  {
    "base": "some-branch",
    "body": "This PR was opened by the [Changesets release](https://github.com/changesets/action) GitHub action. When you're ready to do a release, you can merge this and publish to npm yourself or [setup this action to publish automatically](https://github.com/changesets/action#with-publishing). If you're not ready to do a release yet, that's fine, whenever you add more changesets to some-branch, this PR will be updated.


# Releases

> The changelog information of each package has been omitted from this message, as the content exceeds the size limit.

## simple-project-pkg-a@1.1.0

",
    "head": "changeset-release/some-branch",
    "owner": "changesets",
    "repo": "action",
    "title": "Version Packages",
  },
]
`;

exports[`version doesn't include ignored package that got a dependency update in the PR body 1`] = `
[
  {
    "base": "some-branch",
    "body": "This PR was opened by the [Changesets release](https://github.com/changesets/action) GitHub action. When you're ready to do a release, you can merge this and publish to npm yourself or [setup this action to publish automatically](https://github.com/changesets/action#with-publishing). If you're not ready to do a release yet, that's fine, whenever you add more changesets to some-branch, this PR will be updated.


# Releases
## ignored-package-pkg-b@1.1.0

### Minor Changes

-   Awesome feature
",
    "head": "changeset-release/some-branch",
    "owner": "changesets",
    "repo": "action",
    "title": "Version Packages",
  },
]
`;

exports[`version only includes bumped packages in the PR body 1`] = `
[
  {
    "base": "some-branch",
    "body": "This PR was opened by the [Changesets release](https://github.com/changesets/action) GitHub action. When you're ready to do a release, you can merge this and publish to npm yourself or [setup this action to publish automatically](https://github.com/changesets/action#with-publishing). If you're not ready to do a release yet, that's fine, whenever you add more changesets to some-branch, this PR will be updated.


# Releases
## simple-project-pkg-a@1.1.0

### Minor Changes

-   Awesome feature
",
    "head": "changeset-release/some-branch",
    "owner": "changesets",
    "repo": "action",
    "title": "Version Packages",
  },
]
`;
